index.js

```diff
@@ -1147,49 +1140,27 @@
             );
           }) :
           Z.reduce(function(e, t) {
-            return isEmpty(expType.keys) || isEmpty(t.keys) ?
-              e :
-              Z.chain(function(r) {
-                //  The `a` in `Functor f => f a` corresponds to the `a`
-                //  in `Maybe a` but to the `b` in `Either a b`. A type
-                //  variable's $1 will correspond to either $1 or $2 of
-                //  the actual type depending on the actual type's arity.
-                var offset = t.keys.length - expType.keys.length;
-                return expType.keys.reduce(function(e, k, idx) {
-                  var extractor = t.types[t.keys[offset + idx]].extractor;
-                  var innerValues = Z.chain(extractor, values);
-                  return Z.chain(
-                    function(r) {
-                      return recur(env,
-                                   typeInfo,
-                                   r.typeVarMap,
-                                   expType.types[k].type,
-                                   index,
-                                   Z.concat(propPath, [k]),
-                                   innerValues);
-                    },
-                    Z.reduce(function(e, x) {
-                      var t = expType.types[k].type;
-                      return Z.chain(function(r) {
-                        return test(env)(t)(x) ? Right(r) : Left(function() {
```

Comment:
This was the problem. `test(env)(t)(x)` expands to `satisfactoryTypes(env, {name: 'name', constraints: {}, types: [t]}, {}, t, 0, [], [x]).isRight`, which throws away useful information in the failure case. This forced us to infer lost context below (see the uses of `typeVarConstraintViolation` and `invalidValue`).